src/main.c

```diff
@@ -5,380 +5,80 @@
-		*client_fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
-		if (client_fd <= 0) {
-			printk("socket() failed, errno: %d\n", errno);
-			return -1;
-		}
-
-		hints.ai_socktype = SOCK_STREAM;
-	} else {
-		return -1;
-	}
-
-	err = getaddrinfo(SERVER_HOSTNAME, NULL, &hints, &res);
-	if (err) {
-		printk("getaddrinfo() failed, err %d\n", errno);
-		return -1;
-	}
-
-	if (current_protocol == TIMEOUT_UDP) {
-		((struct sockaddr_in *)res->ai_addr)->sin_port =
-			htons(UDP_PORT);
-	} else if (current_protocol == TIMEOUT_TCP) {
-		((struct sockaddr_in *)res->ai_addr)->sin_port =
-			htons(TCP_PORT);
-	}
-
-	err = connect(*client_fd, res->ai_addr, sizeof(struct sockaddr_in));
-	if (err) {
-		printk("connect failed, errno: %d\n\r", errno);
-		return -1;
-	}
-
-	printk("Connected to server with protocol %s\n",
-	       get_protocol_name(current_protocol));
-
-	return 0;
-}
-
-int get_timeout(bool timed_out)
-{
-	static int udp_counter = 0;
-	static int curr_timeout = 0;
-	static int upper = 0;
-	static int lower = 0;
-	static bool using_binary_search = false;
-
-	// Switch to binary search
-	if (!using_binary_search && timed_out) {
-		using_binary_search = true;
-	}
-
-	if (using_binary_search) {
-		// Adjust bounds
-		if (timed_out) {
-			upper = curr_timeout;
-		} else {
-			lower = curr_timeout;
-		}
 
-		// Found timeout rule
-		if ((upper - lower) == 1) {
-			if (current_protocol == TIMEOUT_UDP) {
-				keep_alive_time.udp = lower;
-				printk("Finished measuring for %s.\nKeep-alive time set at: %d seconds\n",
-				       "UDP", upper);
-			} else if (current_protocol == TIMEOUT_TCP) {
-				keep_alive_time.tcp = lower;
-				printk("Finished measuring for %s.\nKeep-alive time set at: %d seconds\n",
-				       "TCP", upper);
-			}
+		network_status = evt->nw_reg_status;
 
-			curr_timeout = 0;
-			using_binary_search = false;
-		} else {
-			curr_timeout = lower + ((upper - lower) / 2);
-		}
-	} else {
-		lower = curr_timeout;
-		if (current_protocol == TIMEOUT_UDP) {
-			udp_counter++;
-			curr_timeout = udp_counter * udp_counter;
-		} else if (current_protocol == TIMEOUT_TCP) {
-			if (curr_timeout == 0) {
-				curr_timeout = TCP_INITIAL_TIMEOUT;
-			}
-			curr_timeout *= 1.5;
-		}
+		break;
+	default:
+		break;
 	}
-	return curr_timeout;
 }
 
 void main(void)
 {
 	int err;
-	int client_fd;
-	int timeout_s = get_timeout(false);
 
-	printk("TCP client started\n");
+	printk("NAT-test client started\n");
 	printk("Version: %s\n", CONFIG_NAT_TEST_VERSION);
+
 	printk("Setting up LTE connection\n");
 
-	err = lte_lc_init_and_connect();
+	err = lte_lc_init_and_connect_async(lte_handler);
 	if (err) {
-		printk("LTE link could not be established, error: %d\n", err);
+		printk("Error initializing and connecting to LTE, error: %d\n",
+		       err);
 		return;
 	}
 
+	while ((network_status != LTE_LC_NW_REG_REGISTERED_HOME) &&
+	       (network_status != LTE_LC_NW_REG_REGISTERED_ROAMING)) {
+		k_sleep(K_SECONDS(1));
```

Comment:
I'd wait for a sempahore here, and give it in the event handler when one of these events occur

Reply:
I decided against a semaphore here as I could only use it this that one time. Felt weird to have a semaphore which was practically unused after main() is run.

Reply:
I'd go for a semaphore, it's the common approach to such synchronizing between threads/event handlers.

Reply:
Added the semaphore